### PR TITLE
xscreensaver: Add service module and suid wrapper patch

### LINKED ISSUE #130218: xscreensaver warning
**Describe the bug**
When xscreensaver is launched it prints the following warning:
```
xscreensaver-auth: 22:54:53: OOM: /proc/1919/oom_score_adj: Permission denied
xscreensaver-auth: 22:54:53:   To prevent the kernel from randomly unlocking
xscreensaver-auth: 22:54:53:   your screen via the out-of-memory killer,
xscreensaver-auth: 22:54:53:   "xscreensaver-auth" must be setuid root.
```

**Expected behavior**
xscreensaver-auth is wrapped to suid and this wrapper is used by xscreensaver so that it can set oom score.

**Notify maintainers**
@7c6f434c 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.12.14-zen1, NixOS, 21.05.20210713.2a96414 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210601_5985b8b`
 - channels(root): `"nixos-21.05.1486.2a96414d7e3"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by vancluever**
## Description of changes

This adds the NixOS module (from @aidalgol) and a patch for XScreenSaver that ensures that the suid wrapper for xscreensaver-auth is run correctly.

The patch is a simple update to drivers/xscreensaver.c that inserts /run/wrappers/bin before the `DEFAULT_PATH_PREFIX`, which is the directory for xscreensaver hacks/demos, and should be preserved.

The wrapper directory can be modified in the derivation, or even disabled.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Fixes #130218.

### CODE REVIEW

**Comment by vancluever** [comment]
Fixes #130218

**Comment by vancluever** [comment]
@aidalgol I took your NixOS module and tests and expanded upon them :)

You're on here as co-contributor but let me know if you want it amended! I switched maintainer to myself though.

**Comment by vancluever** [comment]
Also re: release notes, I haven't added anything since 23.11 is pretty close and there's no notes for 24.05 yet. But happy to add anything if need be!

**Comment by vancluever** [comment]
Result of  `nix run nixpkgs#nixpkgs-review -- rev HEAD`:

1 package blacklisted:
nixos-install-tools

5 packages built:
caffeine-ng deepin.deepin-screensaver lumina.lumina xscreensaver xsecurelock

**Comment by vancluever** [comment]
Tested with local copy of module here and looks like everything is good.

File: nixos/tests/xscreensaver.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
```
Comment (AndersonTorres) [inline nixos/tests/xscreensaver.nix:1]: Tests should be in a separated commit

File: nixos/modules/services/x11/xscreensaver.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/services/x11/xscreensaver.nix:3]:
```suggestion
```

Do not use global with.

File: pkgs/by-name/xs/xscreensaver/package.nix [hunk 3]
```diff
@@ -26,6 +26,11 @@
 , systemd
 , forceInstallAllHacks ? false
 , withSystemd ? lib.meta.availableOn stdenv.hostPlatform systemd
+, nixosTests
+
+  # This is only included in the rare case it needs to be tuned, or if you want
+  # to exclude it altogether (just set to empty string)
+, withWrapperPrefix ? null
```
Comment (AndersonTorres) [inline pkgs/by-name/xs/xscreensaver/package.nix:33]: Do not use `null`.

File: pkgs/by-name/xs/xscreensaver/package.nix [hunk 4]
```diff
@@ -67,10 +72,17 @@ stdenv.mkDerivation (finalAttrs: {
...
+  patches = [
+    ./xscreensaver-wrapper-prefix.patch
+  ];
+
   preConfigure = ''
     # Fix installation paths for GTK resources.
     sed -e 's%@GTK_DATADIR@%@datadir@% ; s%@PO_DATADIR@%@datadir@%' \
       -i driver/Makefile.in po/Makefile.in.in
+  '' + lib.optionalString (builtins.isString withWrapperPrefix) ''
+    sed -E 's%^(#define NIXOS_WRAPPER_PREFIX).*$%\1 "${withWrapperPrefix}"%' \
```
Comment (AndersonTorres) [inline pkgs/by-name/xs/xscreensaver/package.nix:84]: You should use  `substituteAll` function instead.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/tests/xscreensaver.nix (hunk 1 above)
Comment (vancluever) [reply nixos/tests/xscreensaver.nix:1]: Done! Also split out the package and the module while I was at it.

File: pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ static void
+ hack_environment (void)
+ {
++  static const char *wrapper_path = NIXOS_WRAPPER_PREFIX;
+   static const char *def_path = DEFAULT_PATH_PREFIX;
+   const char *opath = getenv("PATH");
+   char *npath;
+   if (! opath) opath = "/bin:/usr/bin";  /* WTF */
+-  npath = (char *) malloc(strlen(def_path) + strlen(opath) + 20);
++  npath = (char *) malloc(strlen(wrapper_path) + strlen(def_path) + strlen(opath) + 24);
```
[the remark this answers is not in the document]
Comment (7c6f434c) [reply pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:22]: Now I can't help wondering if 20 is _less_ than intended: something with UTF-32 default would need 28 bytes just for «PATH=», «:», and final `\0`. Of course, we all hope this doesn't exist and it was just UTF-16 + a bit of margin…

**Review by 7c6f434c — APPROVED** [review]

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+   strcpy (npath, "PATH=");
++  if (wrapper_path && *wrapper_path)
++    {
```
[the remark this answers is not in the document]
Comment (vancluever) [reply pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:25]: Yep; you can also see the same style in `hack_environment` in [`driver/demo-Gtk.c`](https://github.com/Zygo/xscreensaver/blob/master/driver/demo-Gtk.c#L3711-L3726) which is where this code was derived.

(Note that I didn't patch `demo-Gtk.c` as it seems that it's only used for demos and not `xscreensaver` or `xscreensaver-auth`).

File: pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch (hunk 5 above)
[the remark this answers is not in the document]
Comment (vancluever) [reply pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:22]: Yeah I guess it's never come up as it's in a lot of places in the code (e.g. margins of varying lengths). Given that, someone probably would need to ping jwz about it if it was a major concern :slightly_smiling_face:

**Comment by vancluever** [comment]
Reviewing the build error and making correction.

**Comment by vancluever** [comment]
**Updated** Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>caffeine-ng</li>
    <li>deepin.deepin-screensaver</li>
    <li>lumina.lumina</li>
    <li>xscreensaver</li>
    <li>xscreensaver.man</li>
    <li>xsecurelock</li>
  </ul>
</details>

**Comment by vancluever** [comment]
@AndersonTorres just following up - have you been able to look at this again? Let me know if anything else needs done.

File: pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch:22]: I would add 32 and a comment like "defensive extra room".

**Review by AndersonTorres — APPROVED** [review]
lgtm

**Comment by vancluever** [comment]
:tada: Thanks @AndersonTorres!

File: pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch (hunk 5 above)
[the remark this answers is not in the document]
Comment (vancluever) [reply pkgs/by-name/xs/xscreensaver/xscreensaver-wrapper-prefix.patch]: Done now!

**Comment by AndersonTorres** [comment]
As the latest task,  can you rebase this against master? You were working with Xscreensaver 6.06 whilst the current is 6.08:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/xs/xscreensaver/package.nix

**Comment by vancluever** [comment]
@AndersonTorres done now!

**Comment by AndersonTorres** [comment]
Now squash the commits according to the contributing guidelines.

**Comment by vancluever** [comment]
Thanks for the support in this @AndersonTorres and @7c6f434c!

**Pull request merged**